src/Lean/Elab/Declaration.lean

```diff
@@ -327,35 +326,21 @@ def elabMutual : CommandElab := fun stx => do
     for attrName in toErase do
       Attribute.erase declName attrName
 
-def expandInitCmd (builtin : Bool) : Macro := fun stx => do
-  let optVisibility := stx[0]
-  let optHeader     := stx[2]
-  let doSeq         := stx[3]
-  let attrId        := mkIdentFrom stx <| if builtin then `builtinInit else `init
-  if optHeader.isNone then
-    unless optVisibility.isNone do
-      Macro.throwError "invalid initialization command, 'visibility' modifer is not allowed"
-    `(@[$attrId:ident]def initFn : IO Unit := do $doSeq)
-  else
-    let id   := optHeader[0]
-    let type := optHeader[1][1]
-    if optVisibility.isNone then
-      `(def initFn : IO $type := do $doSeq
-        @[$attrId:ident initFn] opaque $id : $type)
-    else if optVisibility[0].getKind == ``Parser.Command.private then
-      `(def initFn : IO $type := do $doSeq
-        @[$attrId:ident initFn] private opaque $id : $type)
-    else if optVisibility[0].getKind == ``Parser.Command.protected then
+@[builtinMacro Lean.Parser.Command.«initialize»] def expandInitialize : Macro
+  | stx@`($declModifiers:declModifiers $kw:initializeKeyword $[$id? : $type? ←]? $doSeq) => do
+    let attrId := mkIdentFrom stx <| if kw.raw[0].isToken "initialize" then `init else `builtinInit
+    if let (some id, some type) := (id?, type?) then
+      let `(Parser.Command.declModifiersT| $[$doc?:docComment]? $[@[$attrs?,*]]? $(vis?)? $[unsafe%$unsafe?]?) := stx[0]
+        | Macro.throwErrorAt declModifiers "invalid initialization command, unexpected modifiers"
+      let declModifiers ← `(Parser.Command.declModifiersT| $[$doc?:docComment]? @[$attrId:ident initFn, $(attrs?.getD ∅),*] $(vis?)? $[unsafe%$unsafe?]?)
       `(def initFn : IO $type := do $doSeq
-        @[$attrId:ident initFn] protected opaque $id : $type)
+        $(⟨declModifiers⟩):declModifiers opaque $id : $type)
```

Comment:
How is an `unsafe` from `declModifiers` meant to work here? Should it not also apply to the `initFn`?

Reply:
Nice catch, fixed

Reply:
Well, I applied it only to the initializer function now. In theory, these are orthogonal distinctions of "inner" and "outer" unsafety. In practice, `unsafe initialize` is probably always meant to be inner unsafety only. If we had mathlib4's `unsafe` term macro for inner unsafety, we could restore this branch's behavior.